[NFC][testsuite] Add lit 'libdispatch' feature

yln opened the pull request:
I am picking up the effort to make TSan work (better) with Swift on Linux. Currently landing upstream patches for supporting interception of libdispatch on Linux.

The goal of this PR is to identify tests of the dispatch functionality. I am assuming that `REQUIRES: objc_interop` is used as a proxy for libdispatch availability.

The following tests also `import Foundation`. My understanding is that Foundation is not available on Linux. Can you go through the tests and mark the ones which could be modified to work without Foundation, i.e., tests in which Foundation is just used as a utility/helper, but we are not actually testing Dispatch<>Foundation interaction.

- [x] test/ClangImporter/Dispatch_test.swift: YES, removed unnecessary `import Foundation`
- [x] test/PrintAsObjC/dispatch.swift: NO, seems to be real objc_interop test
- [x] test/stdlib/Dispatch.swift: YES, broke out parts that depend on Foundation (`Date`)
- [x] test/stdlib/DispatchData.swift: YES, removed unnecessary `import Foundation`
- [x] test/stdlib/DispatchDeprecationMacOS.swift: NO, deprecation test
- [x] test/stdlib/DispatchDeprecationWatchOS.swift: NO, deprecation test

The list was compiled by searching for the content `objc_interop` in files with name containing `dispatch`. Are there any other tests we should consider here?

I will follow up with PRs to make those tests work without Foundation, actually work on Linux, and then also enable the lit feature flag on Linux. Any help or feedback greatly appreciated!

Previous effort: https://github.com/apple/swift/pull/19836

compnerd commented:
This is definitely a good thing to go through and identify.  `objc_interop` is used as a proxy for many things, including ObjC Interop enabled in the swift runtime, the presence of an objective-c runtime, the availability of an Objective-C foundation implementation, the availability of a foundation implementation (which could be satisfied by swift-corelibs-foundation), and the availability of libdispatch.  Breaking those down into an appropriate feature would improve the coverage of the tests on other platforms.

compnerd commented:
@swift-ci please test

gottesmm commented on test/lit.cfg:
```diff
@@ -633,6 +633,8 @@ if run_vendor == 'apple':
     )
 
     config.available_features.add('objc_interop')
+    config.available_features.add('dispatch')
```
Can you use libdispatch instead of dispatch? That way it is really clear what this is for?

yln commented:
@swift-ci please test

devincoughlin commented on test/stdlib/DispatchDate.swift:
```diff
@@ -0,0 +1,47 @@
+// RUN: %empty-directory(%t)
```
Is it possible to just use

// RUN: %target-run-simple-swift

here?

devincoughlin reviewed (approved):
LGTM with one comment in line. It is great to see finer grained, meaningful lit features for this!

shahmishal commented:
@swift-ci please test

yln committed:
[testsuite] Allow libdispatch tests to run on non-Apple platforms

The lit feature `objc_interop` is used as a proxy for the availability
of libdispatch and Foundation.

Although we don't support Obj-C on non-Apple platforms, we support
libdispatch and Foundation. Let's break up `objc_interop` into
finer-grained categories so we can run more tests on non-Apple
platforms.

This patch adds lit features `libdispatch` and `foundation` (currently
only enabled on Apple platforms) and removes the `objc_interop`
dependency from libdispatch tests.

yln commented:
@swift-ci please test and merge

compnerd commented:
@swift-ci please test and merge

yln commented:
@swift-ci please test and merge

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test macOS platform

yln commented:
@swift-ci please test and merge
